### PR TITLE
[pt1][quant] Add the FP16 weight support for LSTM in dynamic_quantize

### PR DESCRIPTION
**Pull request by jianyuh**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26378 [WIP][pt1][quant] Add the serialization support for FP16 LSTM
* **#25975 [pt1][quant] Add the FP16 weight support for LSTM in dynamic_quantize**

We would like to add the FP16 weight support for the dynamic quantized LSTM.

Todo:
- Deduplicate the code between from_float and __init__.

Differential Revision: [D17299116](https://our.internmc.facebook.com/intern/diff/D17299116/)

### CODE REVIEW

File: torch/nn/quantized/dynamic/modules/rnn.py [hunk 1]
```diff
@@ -58,45 +59,70 @@ def __init__(self, mode, input_size, hidden_size,
...
-                # bias vector is needed in standard definition.
-                b_hh = torch._empty_affine_quantized(
-                    [gate_size], scale=1, zero_point=0, dtype=torch.qint32)
+                    if dtype == torch.qint8:
+                        # for each layer, for each direction we need to quantize and pack
+                        # weights and pack parameters in this order:
+                        #
+                        #   w_ih, w_hh, b_ih, b_hh
+                        packed_weight = \
+                            torch.ops.quantized.linear_prepack(qweight)
```
Comment (jamesr66a) [inline torch/nn/quantized/dynamic/modules/rnn.py]: This should include bias, right? 

It's probably worth it to rebase on top of my change: https://github.com/pytorch/pytorch/pull/25877

File: torch/nn/quantized/dynamic/modules/rnn.py [hunk 2]
```diff
@@ -203,38 +235,56 @@ def from_float(cls, mod):
...
-
-                    params = [packed_weight, bias]
-                    pos_names = ['w', 'b']
-                    ret_name = ['{}_{}_l{}{}'.format(
-                        name, ihhh, layer, suffix) for name in pos_names]
-                    quantized_weights.append(qweight)
-                    packed_weights.append(ret_name[0])
-                    return params, ret_name
+
+                    if dtype == torch.qint8:
```
Comment (jamesr66a) [inline torch/nn/quantized/dynamic/modules/rnn.py]: Same comments as above

BTW we should probably deduplicate this code between from_float and the constructor

**Review by jamesr66a — COMMENTED** [review]
Looks fine on the whole, but I think you should stack on top of my PR since it changes code you're changing pretty significantly

**Review by jamesr66a — CHANGES_REQUESTED** [review]
Cool! Added some comments. In general I think we need to make sure all your fp16 code is using the new weights layout I introduced, as well as figure out that JIT issue you saw

File: torch/nn/quantized/dynamic/modules/rnn.py [hunk 3]
```diff
@@ -51,47 +52,76 @@ def __init__(self, mode, input_size, hidden_size,
...
+                        ret_name = ['{}_{}_l{}{}'.format(
+                            name, ihhh, layer, suffix) for name in pos_names]
+                        return params, ret_name
+                    else:
+                        # for each layer, for each direction we need to quantize and pack
+                        # weights and pack parameters in this order:
+                        #
+                        #   packed_ih, packed_hh, b_ih, b_hh
+                        packed_weight = torch.fbgemm_pack_gemm_matrix_fp16(
+                            qweight.clone().float())
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/nn/quantized/dynamic/modules/rnn.py]: Right: I have removed that.

File: torch/nn/quantized/dynamic/modules/rnn.py [hunk 4]
```diff
@@ -179,23 +209,29 @@ def __setstate__(self, state):
...
-            # import until we need it.
-            from torch.quantization.QConfig import default_dynamic_qconfig
-            weight_observer = default_dynamic_qconfig.weight()
-        assert weight_observer.dtype == torch.qint8, 'Weight observer must have dtype torch.qint8'
+
+        if dtype != torch.qint8 and dtype != torch.float16:
+            raise RuntimeError('Unsupported dtype: {}'.format(dtype))
+
+        if dtype == torch.qint8:
+            # When dtype = torch.float16, we don't need weight_observer
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/nn/quantized/dynamic/modules/rnn.py]: Sure! Have moved it above the if statement.

File: torch/nn/quantized/dynamic/modules/rnn.py (hunk 2 above)
Comment (jianyuh) [reply torch/nn/quantized/dynamic/modules/rnn.py]: Have rebased it. Yes: we should consider a way to deduplicate the code between `from_float` and `__init__`.

File: torch/nn/quantized/dynamic/modules/rnn.py (hunk 1 above)
Comment (jianyuh) [reply torch/nn/quantized/dynamic/modules/rnn.py]: Have rebased on top of your Diff. Right: it should include the bias.

File: torch/nn/quantized/dynamic/modules/rnn.py [hunk 5]
```diff
@@ -51,47 +52,76 @@ def __init__(self, mode, input_size, hidden_size,
...
-                    [gate_size, hidden_size], scale=1, zero_point=0, dtype=torch.qint8)
-                b_ih = torch._empty_affine_quantized(
-                    [gate_size], scale=1, zero_point=0, dtype=torch.qint32)
-                # Second bias vector included for CuDNN compatibility. Only one
-                # bias vector is needed in standard definition.
-                b_hh = torch._empty_affine_quantized(
-                    [gate_size], scale=1, zero_point=0, dtype=torch.qint32)
+                def process_weights(ihhh, layer, suffix, qweight, bias, dtype):
+                    weight_name = 'weight_{}_l{}{}'.format(ihhh, layer, suffix)
+                    bias_name = 'bias_{}_l{}{}'.format(ihhh, layer, suffix)
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/nn/quantized/dynamic/modules/rnn.py]: Have removed them.

File: torch/nn/quantized/dynamic/modules/rnn.py [hunk 6]
```diff
@@ -51,47 +52,76 @@ def __init__(self, mode, input_size, hidden_size,
...
+
+                if dtype == torch.qint8:
+                    for (ih, ih_name), (hh, hh_name) in zip(zip(ih_params, ih_param_names), zip(hh_params, hh_param_names)):
+                        self._all_weight_names.extend([ih_name, hh_name])
+                        self._all_weight_values.extend([ih, hh])
+                else:
+                    for (ih, ih_name), (hh, hh_name) in zip(zip(ih_params, ih_param_names), zip(hh_params, hh_param_names)):
+                        self.register_buffer(ih_name, torch.tensor(
+                            ih) if not isinstance(ih, torch.Tensor) else ih)
+                        self.register_buffer(hh_name, torch.tensor(
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/nn/quantized/dynamic/modules/rnn.py]: Updated

File: torch/quantization/quantize.py [hunk 7]
```diff
@@ -331,7 +332,11 @@ def swap_module(mod, mapping):
     new_mod = mod
     if hasattr(mod, 'qconfig') and mod.qconfig is not None:
         if type(mod) in mapping:
-            new_mod = mapping[type(mod)].from_float(mod)
+            if (dtype == torch.qint8):
+                new_mod = mapping[type(mod)].from_float(mod)
+            else:  # dtype == torch.float16
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/quantization/quantize.py]: Updated.

File: test/test_quantization.py [hunk 8]
```diff
@@ -490,12 +490,20 @@ def test_quantized_rnn(self):
             torch.nn.LSTM: torch.nn.quantized.dynamic.LSTM,
         }
         model_int8 = quantize_dynamic(
-            model, qconfig_dynamic_dict, default_dynamic_module_mapping
+            model=model, qconfig_dict=qconfig_dynamic_dict, mapping=default_dynamic_module_mapping,
+            dtype=torch.torch.qint8
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply test/test_quantization.py]: Fixed.

File: test/test_quantization.py [hunk 9]
```diff
@@ -538,15 +546,22 @@ def forward(self, x, hiddens):
...
-
+        # Compare fp16 quantized to unquantized
+        output_fp16, final_hiddens_fp16 = cell_fp16(x, hiddens)
+
+        torch.testing.assert_allclose(output_fp16, ref_out)
+        self.assertEqual(output_fp16, ref_out)
+        for out, ref in zip(final_hiddens_fp16, ref_hid):
+            torch.testing.assert_allclose(out, ref)
+
+#         # Test save/load
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply test/test_quantization.py]: For torch.float, we don’t need `dynamic_val`. But we need to keep the original_weight for serialization. This caused some inconsistency between qint8 and float16 serialization.

[the remark this answers is not in the document]
Comment (jamesr66a) [reply test/test_quantization.py]: @jianyuh can you just make a convention to use dynamic_val to hold original_weight as a single element under fp16? Then in __setstate__ we would just branch on dtype and either pull out the dynamic vals normally (int8) or pull assert dynamic val has 1 element then just pull that out and assign to original_weight

Btw you should put the serialization testing code back

[the remark this answers is not in the document]
Comment (jianyuh) [reply test/test_quantization.py]: Right: not removing the serialization testing code (the int8 serialization code is kept as it is). For the FP16 serialization, I will separate as a new PR: https://github.com/pytorch/pytorch/pull/26378 (Still WIP).

**Review by jamesr66a — APPROVED** [review]
I think this is good to go. Only minor comments

File: torch/nn/quantized/dynamic/modules/rnn.py [hunk 10]
```diff
@@ -173,29 +194,36 @@ def __setstate__(self, state):
...
-            weight_observer = mod.qconfig.weight()
-        else:
-            # We have the circular import issues if we import the qconfig in the beginning of this file:
-            # https://github.com/pytorch/pytorch/pull/24231. The current workaround is to postpone the
-            # import until we need it.
-            from torch.quantization.QConfig import default_dynamic_qconfig
-            weight_observer = default_dynamic_qconfig.weight()
-        assert weight_observer.dtype == torch.qint8, 'Weight observer must have dtype torch.qint8'
+
+        if dtype != torch.qint8 and dtype != torch.float16:
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/nn/quantized/dynamic/modules/rnn.py]: Fixed.

File: torch/nn/quantized/dynamic/modules/rnn.py [hunk 11]
```diff
@@ -173,29 +194,36 @@ def __setstate__(self, state):
...
+            raise RuntimeError('Unsupported dtype: {}'.format(dtype))
+
+        # When dtype = torch.float16, we don't need weight_observer
+        if dtype == torch.qint8:
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/nn/quantized/dynamic/modules/rnn.py]: Fixed.

File: torch/quantization/quantize.py [hunk 12]
```diff
@@ -331,7 +332,13 @@ def swap_module(mod, mapping):
     new_mod = mod
     if hasattr(mod, 'qconfig') and mod.qconfig is not None:
         if type(mod) in mapping:
-            new_mod = mapping[type(mod)].from_float(mod)
+            if dtype != torch.qint8 and dtype != torch.float16:
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/quantization/quantize.py]: Fixed.

File: torch/quantization/quantize.py [hunk 13]
```diff
@@ -331,7 +332,13 @@ def swap_module(mod, mapping):
...
+                 raise RuntimeError('Unsupported dtype: {}'.format(dtype))
+            if (dtype == torch.qint8):
```
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/quantization/quantize.py]: Fixed.

File: torch/quantization/quantize.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jianyuh) [reply torch/quantization/quantize.py]: Fixed. Should we only keep one copy of `supported_scalar_types` globally? Currently we need to use it in `torch/quantization/quantize.py` and `torch/nn/quantized/dynamic/modules/rnn.py`.

**Pull request closed without merging**